Add admin panel for actuator information

atarix83 opened the pull request:
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #801 
* Requires DSpace/DSpace#8166 (if a REST API PR is required to test this)

## Description
This PR introduce a simple admin panel to display the "info" and "health" actuator responses in a nice manner.

## Instructions for Reviewers
Try using the REST branch available in the [PR](https://github.com/DSpace/DSpace/pull/8166). Check the new admin panel available (route `/health`). Check the new route `/app/health` returns the server status properly.

List of changes in this PR:
* Added a new route `/app/health` (available only with SSR) that return a json with the server status retrieved by using `/actuator/health` endpoint
* Added a new entry to the admin sidebar menu
* Created a new page (`/health`) for displaying the panel with actuator information
![Schermata da 2022-05-03 15-36-34](https://user-images.githubusercontent.com/2486489/166463556-928fd74d-f884-468a-8691-911115b56cdc.png)
![Schermata da 2022-05-03 15-36-45](https://user-images.githubusercontent.com/2486489/166463581-f51bf712-1d04-4eb2-9ce8-b241da53a4ec.png)



**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

artlowel commented on src/app/health-page/health-data.service.ts:
```diff
@@ -0,0 +1,32 @@
+@Injectable({
+    providedIn: 'root'
+})
+export class HealthDataService {
```
I wouldn't call it HealthDataService if it doesn't have the `@dataService` decorator. To prevent confusion. 

Or even better, actually implement it as a dataservice. I can see these calls are very straightforward, so it would be a little overhead for this particular case, but the advantage is consistency.

Every other request the UI makes to our backend can be found in the request section of the ngrx store, except for this service. So at some point in the future someone will be debugging some problem and waste a bunch of time wondering why they see requests in the network tab that don't show up in the store.

atarix83 replied:
I agree we should avoid to not extending data service, but the main reason why we can't implement the data service is that the rest response come from a spring boot implementation and they don't have the `type` attribute. So we are not able to use the dataservice

artlowel replied:
I see. In that case I'd just rename it to prevent confusion.

artlowel reviewed (changes requested):
Thanks @atarix83!

Currently the "Geoip" section has no options, but it's still possible to click it open. That makes it look as though there is supposed to be content, but it failed to load. 


The following are suggestions to improve the usability, but shouldn't hold back the merging of this PR:

- The top level links (with the checkmarks) are in effect giant buttons. you can click anywhere between the text and the checkmark. But they look like links, that would only work if you click directly on the text. I think it would work better if you used a [[bootstrap card](https://getbootstrap.com/docs/4.6/components/card/#header-and-footer)](https://getbootstrap.com/docs/4.6/components/card/#header-and-footer) for each section, with that top button as the header bar. Or perhaps the [[toggle panels in ng-bootstrap](https://ng-bootstrap.github.io/#/components/accordion/examples#toggle)](https://ng-bootstrap.github.io/#/components/accordion/examples#toggle)
- The plus and minus icons at the top level are too big. That makes it look less polished in my opnion. I also think chevrons, [[as used by those toggle panels](https://ng-bootstrap.github.io/#/components/accordion/examples#toggle)](https://ng-bootstrap.github.io/#/components/accordion/examples#toggle) work better to indicate that something will open when you click it. 
- I'd get rid of the ability to collapse anything below the top level. If I click open the DB section I don't want to have to click open each subsection manually afterwards. I don't see a reason to ever collapse those subsections.

tdonohue reviewed (changes requested):
@atarix83 : Thanks for this!  Overall this works well, but I do think it has some usability & accessibility issues.

1. First, I agree with @artlowel 's feedback.  I think the +/- buttons are way too large, and I agree that clicking at the top level should open all information (rather than having many different subsections to also expand)
2. In the Admin side menu, if you hover over the Health section, the title is "menu.section.icon.health".  It looks like we are missing an i18n key.
3. An accessibility scan shows several small issues with the /health page
    * It has no H1 heading (there's just a list of tabs). This page should likely have a title that says "Health Checks" or similar.
    * There are invalid `aria-control` values. The section headings (e.g. Db, Geoip, etc) seem to have `aria-control=collapseExample` which isn't valid. If we can fix this we should.
    * The tablist has a `role=tablist` but the tabs don't have `role=tab` to match it.  I think you mislabeled the `<a>` tag with `role=tab`, when it should be the `<li>` tag.
4. The icons are unclear / not very useful. The triangle "!" icon has no explanation or meaning to users (e.g. is this an error? Is it a warning? what does it mean?).  Where ever that "!" icon appears, there should be a clear explanation like "Possible issues detected" (or a hover over)
    * Similarly there's no explanation of the green checkbox.  I'd expect there to be either a hover over or when you expand that section it might explain "Status: UP" or similar.
5. On the "Status" tab, the expandable sections are confusingly named (e.g. will people know what "Solrauthoritycore" means?).  Maybe we should use i18n keys to give the sections better names?  For example the name "Solrauthoritycore" might be translated to "Solr : Authority Core". If we use i18n then people could easily give it a different name or translate as needed.
6. On the "Status" tab, in the "Solr" sections, the expandable information is not very useful.  For example, I don't know what "Status: 0" and "Detectedpathtype : particular core" means for a Solr endpoint.  Maybe we should just return a simple message saying "Status: UP" or "Status: DOWN" depending on the status. (And we could just not display "detectedpathtype" whatever that means)
    * This same approach could be used for the DB and GeoIP sections. In the DB section, the value "Validationquery: isValid()" is not useful...it could instead say "Status: UP" or "Status: Responding" or similar.
7. On the "Info" tab, the "Version: 7.3-SNAPSHOT" has no context.  I know this is the DSpace Version, but it's the only part that doesn't have a heading.  Maybe it should be moved into the "App" section where it would make more sense? ![Info](https://user-images.githubusercontent.com/483997/167006170-8b0381ce-c942-4f17-aae8-4653c36d564f.png)
8. On the "Info" tab, the "App" section should likely be renamed to say "Application Backend" or "DSpace Backend".  These settings all come from the backend, but "App" implies they could be coming from the frontend.
9. The "/health" page doesn't work if the overall status is "DOWN". In that scenario the `/actuator/health` endpoint on the backend will return a 503 error. This causes the "/health" page to fail to load, even though the `/actuator/health` endpoint will return useful information about what service is down. 
    * The easiest way to reproduce this is to stop Solr and access this `/health` page.  With Solr down, it is possible to still login & go to `/health`, but the `/health` page will fail to load.

Beyond that, the basics all work.  I just think the user experience is a bit messy right now.  If we clean this up a bit per @artlowel and my feedback, it'd be great!

atarix83 commented:
@artlowel  @tdonohue 

I've changed the interface by addressing your feedback now it looks like the following
![Schermata da 2022-05-16 19-20-59](https://user-images.githubusercontent.com/2486489/168648449-e3c7d069-db05-4bb2-8f47-5148a626ef06.png)

In details : 
> In the Admin side menu, if you hover over the Health section, the title is "menu.section.icon.health". It looks like we are missing an i18n key.

fixed

> An accessibility scan shows several small issues with the /health page
    It has no H1 heading (there's just a list of tabs). This page should likely have a title that says "Health Checks" or similar.
    There are invalid aria-control values. The section headings (e.g. Db, Geoip, etc) seem to have aria-control=collapseExample which isn't valid. If we can fix this we should.
    The tablist has a role=tablist but the tabs don't have role=tab to match it. I think you mislabeled the <a> tag with role=tab, when it should be the <li> tag.

those errors come from the ngb-boostrap components we're using, honestly i don't know how to fix 

> The icons are unclear / not very useful. The triangle "!" icon has no explanation or meaning to users (e.g. is this an error? Is it a warning? what does it mean?). Where ever that "!" icon appears, there should be a clear explanation like "Possible issues detected" (or a hover over)
    Similarly there's no explanation of the green checkbox. I'd expect there to be either a hover over or when you expand that section it might explain "Status: UP" or similar.

Added a tooltip for the stats icons

> On the "Status" tab, the expandable sections are confusingly named (e.g. will people know what "Solrauthoritycore" means?). Maybe we should use i18n keys to give the sections better names? For example the name "Solrauthoritycore" might be translated to "Solr : Authority Core". If we use i18n then people could easily give it a different name or translate as needed.

done

> On the "Status" tab, in the "Solr" sections, the expandable information is not very useful. For example, I don't know what "Status: 0" and "Detectedpathtype : particular core" means for a Solr endpoint. Maybe we should just return a simple message saying "Status: UP" or "Status: DOWN" depending on the status. (And we could just not display "detectedpathtype" whatever that means)
    This same approach could be used for the DB and GeoIP sections. In the DB section, the value "Validationquery: isValid()" is not useful...it could instead say "Status: UP" or "Status: Responding" or similar.
    NOTE: I do realize that these messages come from the backend. So, if you'd rather display anything that the backend sends, that's OK. But, I'd therefore also recommend adding the "Status: UP/DOWN" to these sections (as it's currently not displayed). So, for example, a Solr section might show:
        "Status: UP"
        "Status Code: 0"
        "Detected Path Type : particular core"

I think the status information is already given by the icon present in the panel header. Probably we should see if the property `status` can be returned as `status code` by the rest. 
![Schermata da 2022-05-16 19-33-11](https://user-images.githubusercontent.com/2486489/168650361-54f1a802-848f-46d7-927c-52496b551590.png)

> The "/health" page doesn't work if the overall status is "DOWN". In that scenario the /actuator/health endpoint on the backend will return a 503 error. This causes the "/health" page to fail to load, even though the /actuator/health endpoint will return useful information about what service is down.
    The easiest way to reproduce this is to stop Solr and access this /health page. With Solr down, it is possible to still login & go to /health, but the /health page will fail to load.

The problem here is that the HttpClient service thrown an error because of the 503 status. For the moment I've added a fallback message in case an error is thrown. Probably the rest should return a 200 status instead of 503. We can see if there is something that can be fixed on rest side

![Schermata da 2022-05-16 19-20-07](https://user-images.githubusercontent.com/2486489/168651116-4c4e6f2d-a514-43b1-8cd1-ff428591b4ca.png)

tdonohue reviewed (changes requested):
@atarix83 : This looks **much better**.   I've verified that the majority of my initial feedback has been addressed, and that the page works when Solr is down (and it correctly now shows a "Problems detected" status with individual errors from Solr cores).

That said, I see a few things were not yet addressed:

1. (Required) The new Health page needs a header  / page title (as this exists on every other page & are required for accessibility).  I'd recommend the following:
    * First, we need to ensure the page `<title>` exists. Currently, this page just reuses the same title as the page you were on previously.  It should have a title of `DSpace Angular :: Health` or similar
    * Second, this page needs a heading (`<h1>` or `<h2>`) *above the tabs* like all other pages.  I'd recommend the heading say "Health".  and it must exist **above** the current tabs.
    * I'd then recommend we rename the "Health" tab to say "Status", as it's really the status section of the Health page.  You might consider whether to make the "Status" heading (representing the overall status) on that tab slightly smaller... but it is important information obviously.
2. (Required) Unit tests are currently failing. They need to obviously be fixed.
3. (Optional) I noticed you decided to keep the many open/close subsections in the page.  Both @artlowel and I noted that it'd be nice to remove the subsections (e.g. opening the "Database" section should display all the info about the "datasource" and "dspacedatasource".... currently you have to open each individually.  Same with opening the "Java" section on the "info" tab...it's odd to have to individually open the "Runtime" vs "JVM" subsections).
    * I flagged this as optional as I'm OK with leaving this as-is for now if it's too complex.  I just feel it requires a lot of unnecessary clicking to find the information oy may be looking for.

atarix83 commented:
@tdonohue  

feedback addressed

regarding the point 3, I've forgotten to made the change you and Art have suggested. Now the sub-section are opened by default. I'd leave the possibility to collapse if you agree

tdonohue reviewed (approved):
👍 Thanks @atarix83 !  This all looks good to me now & it works well.  At the very least this is a very good initial control panel & we can always enhance/improve things in the future as needed.

@artlowel : When you get a chance could you give this a re-review?  If it looks good to you, we likely can merge this (along with the backend PR which already is at +2)

artlowel commented:
Thanks @atarix83!

I re-tested, and I see that most of my feedback has been addressed, but the issue with the empty Geoip panel remains.

It looks just like any other panel when you open the page: with a chevron to click for more details
![image](https://user-images.githubusercontent.com/1567693/170070024-8d293391-8e95-4996-8de2-660338e1f22f.png)

But the rest response doesn't have details, so if you open it, it simply shows an empty panel
![image](https://user-images.githubusercontent.com/1567693/170070223-48f2a248-2916-4c28-9b96-b7ab5ca85c73.png)

I expect users will think something went wrong. So if at all possible, I'd hide or disable the chevron for sections that don't have details. 

And if there's really no other way, at least put some text in the panel when there's nothing to show: e.g. : "No further details available". That would indicate to users that it's working as expected

davide-negretti commented:
Message "no issue detected" is now shown if both components and details are missing

![image](https://user-images.githubusercontent.com/87638927/172185863-12a6f7dd-0315-4741-81a4-8373b2ed95a2.png)

artlowel commented on src/app/health-page/health-info/health-info-component/health-info-component.component.html:
```diff
@@ -0,0 +1,56 @@
+</div>
+
+
+<!--
+
+  <div class="mb-3 border-bottom" >
+    <div class="w-100 d-flex justify-content-between py-2 mb-0" (click)="collapse.toggle()">
+      <button type="button" class="btn btn-link p-0 " (click)="$event.preventDefault()" [attr.aria-expanded]="!collapse.collapsed"
+              aria-controls="collapseExample">
+        <span [class.h4]="!isNested">{{ healthInfoComponentName | titlecase }}</span>
+      </button>
+      <div class="d-inline-block">
+        <span *ngIf="collapse.collapsed" class="fas fa-chevron-down"></span>
+        <span *ngIf="!collapse.collapsed" class="fas fa-chevron-up"></span>
+      </div>
+    </div>
+    <div #collapse="ngbCollapse" [ngbCollapse]="isCollapsed">
+      <div class="card border-0">
+        <div class="card-body">
+          <ds-health-info-component *ngFor="let entry of healthInfoComponent | dsObjNgFor"
+                                    [healthInfoComponent]="entry.value"
+                                    [healthInfoComponentName]="entry.key"
+                                    [isNested]="true"
+                                    data-test="component"></ds-health-info-component>
+        </div>
+      </div>
+    </div>
+  </div>
+</ng-container>-->
```
This commented out code block can probably be removed

artlowel reviewed (changes requested):
Thanks @atarix83! The UI looks good now.

A few more things to clean up in the code though: there are multiple components in this PR that are called …ComponentComponent: HealthComponentComponent, HealthInfoComponentComponent. That seems confusing. I'd rename them, perhaps something like …ComponentWrapper, …ComponentLoader, …ComponentList

There are also a few blocks of commented out code that can likely be removed

davide-negretti commented:
> A few more things to clean up in the code though: there are multiple components in this PR that are called …ComponentComponent: HealthComponentComponent, HealthInfoComponentComponent. That seems confusing. I'd rename them, perhaps something like …ComponentWrapper, …ComponentLoader, …ComponentList

@artlowel these Angular components contain "component" in their name because they represent an object that is called "component", as you can see in the response of the `/health` endpoint:

![image](https://user-images.githubusercontent.com/87638927/172388102-6ac7683a-49b2-4847-9ece-f97068b568c2.png)

I understand that the name could be confusing, but the names are consistent, e.g. the name of the `health-component` component representing the "component" object is consistent with the name of the `health-status` component representing the "status" object

artlowel commented:
>> A few more things to clean up in the code though: there are multiple components in this PR that are called …ComponentComponent: HealthComponentComponent, HealthInfoComponentComponent. That seems confusing. I'd rename them, perhaps something like …ComponentWrapper, …ComponentLoader, …ComponentList
>
>@artlowel these Angular components contain "component" in their name because they represent an object that is called "component", as you can see in the response of the `/health` endpoint:
>
>![image](https://user-images.githubusercontent.com/87638927/172388102-6ac7683a-49b2-4847-9ece-f97068b568c2.png)
>
>I understand that the name could be confusing, but the names are consistent, e.g. the name of the `health-component` component representing the "component" object is consistent with the name of the `health-status` component representing the "status" object

@davide-negretti In that case can you please add that clarification to the typedocs for those components? I expect other people will be confused about it as well.

davide-negretti commented:
@artlowel I added the TypeDocs

artlowel commented on src/app/health-page/health-panel/health-component/health-component.component.ts:
```diff
@@ -4,6 +4,9 @@ import { HealthComponent } from '../../models/health-component.model';
 import { TranslateService } from '@ngx-translate/core';
 import { AlertType } from '../../../shared/alert/aletr-type';
 
+/**
+ * Show a health component object
+ */
```
That doesn't clarify very much. I was thinking something along the lines of:

```suggestion
  /**
   * A component to render a "health component" object.
   *
   * Note that the word "component" in "health component" doesn't refer to Angular's use of the term
   * but rather to the components used in the response of the health endpoint of Spring's Actuator 
   * API.
   */
```

I'd do something similar for HealthInfoComponentComponent, as well as the two models themselves.

atarix83 replied:
done

artlowel reviewed (approved):
Thanks

tdonohue commented:
Merging as this is now at +2 (as is the backend PR).  Thanks again @atarix83 and @davide-negretti !
